### DISCUSSION TITLE
Recommended handling for large modularized projects with simple extensibility or design patterns

### DISCUSSION THREAD

**Discussion by FSharpCSharp** [AvaloniaUI/Avalonia#6791] [Q&A]
Hello all,

since I can only find very sparse information on this topic, I'll open a discussion here.

My concern is the following: I am planning to use Avalonia in larger projects. I originally come from the WPF world. Here I have implemented large projects with the Prism framework from Microsoft, among others. The advantage here was that I could divide my product into modules to significantly increase the maintainability of the code and thus the entire product.

Now I found a community project to use Prism in Avalonia as well. But this seems to be treated stepmotherly, and is based on a relatively old version of Prism. New important functions like a DialogService or similar is still possible with it. Also some features are only partially or not at all supported.

Some of them I could certainly implement from scratch, but of course it would be a lot of work, because I would have to implement features that already exist in WPF for Avalonia. In the WPF version of Prism, all this would already be there, and the code would be maintained and updated by experts. A separate solution in this area would be good on the one hand, but bad on the other hand, because I would be constantly forced to provide improvements and corrections to the Prism framework itself.

Therefore I wanted to ask here whether and how to deal with larger projects currently in Avalonia. I'm kind of afraid to put everything on this community-driven Prism framework now, that it will be discontinued bit by bit. And then I'm back at the beginning with components that are neither maintained nor usable.

Are there any best practices for this, or any templates on how to handle larger modularized projects in Avalonia?

**Comment by fforjan** [comment]
I'm surprise no one has answered here, but from my experience, I will first be looking at which dependency injection I could use with Avalonia and how decoupled it could be. This would be the step 0 for me to start it.
The next step would be looking at a MDI type framework working with Avalonia and see how it could be fitting then.

It seems AvalonStudio and its AvalonStudio.Shell would be an good way to check and see how they solved those 2 issues ?https://github.com/VitalElement/AvalonStudio

**Reply by MikeCodesDotNET** [reply]
![image](https://user-images.githubusercontent.com/552074/139473978-8b5e1d4c-b8eb-4472-9ad7-5c4b6aa4edea.png)

I've been implementing an MDI control for Avalonia as you can see above. It's really not too difficult to build out this type of functionality, especially if you've done it before with WPF.

**Reply by FSharpCSharp** [reply]
This looks very promising indeed!  That's exactly what I meant, a kind of foundation on which you can build.

My only fear at the moment is that I invest too much in the development of a program with Avalonia, whose APIs are not yet finally fixed, and thus I develop everything twice. In principle, you can build anything, of course, that's clear to me, but the investment to constantly maintain it is quite large for me for a product in beta stage.

Are you planning more tools or toolkits that make life easier? I think if there would be something like that as standard in Avalonia, then that could be a gamechanger. Avalonia has already eliminated some of the crutches of WPF anyway and it would be good if the available base is ready, so to speak, so that you can quickly get to the result in the sense of RAD.

**Reply by msdobrescu** [reply]
Hi, is the above MDI functionality available after all?
I need to assess if this is possible and, if so, how.

**Comment by FSharpCSharp** [comment]
Thank you for the short notice. There are definitely some very interesting projects related to Avalonia. I have already looked at the AvalonStudio. Theoretically, some of it would serve as a basis, but the licensing of the whole thing seems a bit hazy to me. The program itself is released under GNU GPL 3. But the AvalonStudio.Shell you mention somehow isn't, and only comes with the following notice:

> (c) VitalElement 2018
> 
> We are currently working on a proper software license.
> 
> However currently we have the following rules:
> 
> This software is free for non-commercial open source projects.
> 
> This software may not be used in an application that could be considered an IDE.
> 
> Please contact james@vitalelement.co.uk for queries about licensing.
> 
> But anyway there is a license file of type GNU GPL in the directory too.

I think it is this license that makes the code not usable for the majority of applications. If this would be under an Apache or MIT license, then all this would not be so bad. But the biggest obstacle here are some infelicities. 

I was actually hoping that one of the people responsible for the project would get back to me. These topics are basically not new. The point is for me one of the biggest current problems of the project that it is not currently well possible to map larger projects that are modularized without the use of a lot of manpower.

**Reply by danwalmsley** [reply]
thats and old file accidentally left there....

https://github.com/VitalElement/AvalonStudio.Shell/blob/master/LICENSE

is the license.

**Reply by FSharpCSharp** [reply]
Yes exactly my opinion. You don't know exactly what you're getting, and the GPL license is all well and good, but from my point of view it's very restrictive for the development of proprietary applications. On the one hand it is good, but there are also clear disadvantages. Microsoft themselves release all their libraries and tools under the MIT license because they have found out that any other licenses are just not good for proprietary use.
